Decrease initial step size in TOV integrator, switch conformal factor interpolant to CubicSpline

nilsvu opened the pull request:
## Proposed changes

Rebasing on #4111 led to strong Hamiltonian constraint violations near the origin in #3903. I tracked this down to the switch from BarycentricRational to CubicSpline for the M(r) and ln(h) interpolators in the TOV solution. It looks like the initial step size in the TOV integrator was too large, leading to significant differences between the two interpolations at the start of the integration interval. Here's a quick plot that shows the large step size near r=0 (dots are integration steps, line is the interpolation):

<img width="315" alt="0c712d5b-bf32-4d92-80a7-2d928267ed37" src="https://user-images.githubusercontent.com/746230/208803290-e5ca1d88-7fb3-4541-9c43-d803b51c81b8.png">

Here's a plot of the adaptive step size in both r and ln(h):

<img width="268" alt="fe635671-3da3-4bd9-9339-16f3c701e2be" src="https://user-images.githubusercontent.com/746230/208803846-29b6437e-fd33-4298-a841-acf3b632f45e.png">

Therefore, I reduced the initial step size to `relative_tolerance * (ln(h_R) - ln(h_C))`. That seems to sample the origin densely enough so that the Hamiltonian constraint violations vanish. Here are the updated plots:

<img width="315" alt="7f14e704-887a-4a40-b12c-ee6a7dce700f" src="https://user-images.githubusercontent.com/746230/208803956-ba255a7e-faf6-448f-af28-196fe0b44710.png">

<img width="273" alt="ac49ea31-f936-4ce1-a707-1c387499bdb1" src="https://user-images.githubusercontent.com/746230/208803985-4211c4b4-92a8-482e-9848-a9c20fad2886.png">

Any other ideas would be welcome too.

With the dense sampling near the origin the BarycentricRational interpolation for the conformal factor (which wasn't switched in #4111) became unstable, so I switched it to CubicSpline too.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Bind TOV interpolants in Python

nilsvu committed:
Switch TOV conformal factor interpolant to CubicSpline

BarycentricRational interpolation oscillates strongly when the
step size becomes small.

teukolsky commented:
I think "time step" should be "stepsize" in several places. The integration is in r.

nilsvu commented:
@teukolsky definitely, I'll change the wording. The integration is actually in `ln(h)` :)

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/Tov.cpp:
```diff
@@ -177,11 +177,9 @@ void TovSolution::integrate(
-  if constexpr (CoordSystem == TovCoordinates::Isotropic) {
-    initial_step = -std::min(std::abs(initial_step), std::abs(1.0 / dvars[2]));
-  }
+  const double initial_step =
```
Can you please document why you made this change and the motivation? The previous definition is from Hairer and is what is canonically recommended as the initial step size. Yours is quite arbitrary as far as I can tell.

nilsvu replied:
I dropped this change and instead tightened the absolute tolerance. I'm a bit worried about the tolerances here.. they seem rather fragile. An absolute tolerance of 1e-14 wasn't enough for the XCTS constraints. Also, loosening the relative tolerance to just 1e-12 fails the `test_tov_dp_dr`. And, when tightening the absolute tolerance further (to, say, 1e-30 or even 0), the `test_tov` times out in the Newtonian limit case.

nilsdeppe reviewed (changes requested):
This seems fine, but I wonder if this could be fixed by adjusting the absolute tolerance. The quantities being integrated (u=r^2 & v=m/r~r) are both zero at the origin, I think. I don't know how quickly they grow above the absolute tolerance, but it might be worth adjusting the absolute tolerance to see if that's sufficient of a fix. The fact that decreasing the step size near the origin improved things to me says it's a tolerance issue

nilsvu committed:
Add more captures to Test_Tov.cpp

nilsvu committed:
Tighten absolute tolerance in TOV solver

This resolves an issue where the step size of the integration
is too large initially, which led to significant constraint violations
in the XCTS system for an isotropic TOV star.
